tools/depcheck/pkg/graph/graph.go

```
@@ -55,7 +59,99 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
 	return nil
 }
 
+// RemoveNode deletes the given node from the graph,
+// removing all of its outbound and inbound edges as well.
+func (g *MutableDirectedGraph) RemoveNode(n *Node) {
+	delete(g.nodesByName, n.UniqueName)
+	g.DirectedGraph.RemoveNode(n)
+}
+
 func (g *MutableDirectedGraph) NodeByName(name string) (graph.Node, bool) {
 	n, exists := g.nodesByName[name]
 	return n, exists && g.DirectedGraph.Has(n)
 }
+
+// PruneOrphans recursively removes nodes with no inbound edges.
+// Receives a list of UniqueNames of nodes to ignore.
+// Returns a list of recursively pruned nodes.
+func (g *MutableDirectedGraph) PruneOrphans(originatingNodes []string) []*Node {
+	removed := []*Node{}
+
+	for _, n := range g.Nodes() {
+		node, ok := n.(*Node)
+		if !ok {
+			continue
+		}
+		if len(g.To(n)) > 0 {
+			continue
+		}
+		if contains(node.UniqueName, originatingNodes) {
+			continue
+		}
+
+		g.RemoveNode(node)
+		removed = append(removed, node)
+	}
+
+	if len(removed) == 0 {
+		return []*Node{}
+	}
+
+	return append(removed, g.PruneOrphans(originatingNodes)...)
+}
+
+func contains(needle string, haystack []string) bool {
+	for _, str := range haystack {
+		if needle == str {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Copy creates a new graph instance, preserving all of the nodes and edges
+// from the original graph. The nodes themselves are shallow copies from the
+// original graph.
+func (g *MutableDirectedGraph) Copy() *MutableDirectedGraph {
+	newGraph := NewMutableDirectedGraph()
+
+	// copy nodes
+	for _, n := range g.Nodes() {
+		newNode, ok := n.(*Node)
+		if !ok {
+			continue
+		}
+
+		// ignore error: the only error that could happen is if a name collision occurs.
```

panic instead of ignore